pkg/ruler/api.go

```diff
@@ -268,3 +282,274 @@ func (r *Ruler) alerts(w http.ResponseWriter, req *http.Request) {
 	}
 }
+
+var (
+	// ErrNoNamespace signals the requested namespace does not exist
+	ErrNoNamespace = errors.New("a namespace must be provided in the url")
+	// ErrNoGroupName signals a group name url parameter was not found
+	ErrNoGroupName = errors.New("a matching group name must be provided in the url")
+	// ErrNoRuleGroups signals the rule group requested does not exist
+	ErrNoRuleGroups = errors.New("no rule groups found")
+	// ErrNoUserID is returned when no user ID is provided
+	ErrNoUserID = errors.New("no id provided")
+	// ErrBadRuleGroup is returned when the provided rule group can not be unmarshalled
+	ErrBadRuleGroup = errors.New("unable to decoded rule group")
+)
+
+// ValidateRuleGroup validates a rulegroup
+func ValidateRuleGroup(g rulefmt.RuleGroup) []error {
+	var errs []error
+	for i, r := range g.Rules {
+		for _, err := range r.Validate() {
+			var ruleName string
+			if r.Alert != "" {
+				ruleName = r.Alert
+			} else {
+				ruleName = r.Record
+			}
+			errs = append(errs, &rulefmt.Error{
+				Group:    g.Name,
+				Rule:     i,
+				RuleName: ruleName,
+				Err:      err,
+			})
+		}
+	}
+
+	return errs
+}
+
+// parseNamespace parses the namespace from the provided set of params, in this
+// api these params are derived from the url path
+func parseNamespace(params map[string]string) (string, error) {
+	namespace, exists := params["namespace"]
+	if !exists {
+		return "", ErrNoNamespace
+	}
+
+	namespace, err := url.PathUnescape(namespace) // namespaces needs to be unescaped if in the URL
+	if err != nil {
+		return "", err
+	}
+
+	return namespace, nil
+}
+
+// parseGroupName parses the group name from the provided set of params, in this
+// api these params are derived from the url path
+func parseGroupName(params map[string]string) (string, error) {
+	groupName, exists := params["groupName"]
+	if !exists {
+		return "", ErrNoGroupName
+	}
+
+	groupName, err := url.PathUnescape(groupName) // groupName needs to be unescaped if in the URL
+	if err != nil {
+		return "", err
+	}
+
+	return groupName, nil
+}
+
+func (r *Ruler) listRules(w http.ResponseWriter, req *http.Request) {
+	logger := util.WithContext(req.Context(), util.Logger)
+	userID, err := user.ExtractOrgID(req.Context())
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	vars := mux.Vars(req)
+
+	// Parse the namespace from the url path parameters and return
+	// a 400 if it is invalid and return a full set of rules if no
+	// namespace is provided
+	namespace, err := parseNamespace(vars)
+	if err != nil {
+		// If a namespace is not provided continue as usual and return a full set of rules
+		if err != ErrNoNamespace {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+		level.Debug(logger).Log("msg", "retrieving rule groups with namespace", "userID", userID, "namespace", namespace)
+	}
+
+	level.Debug(logger).Log("msg", "retrieving rule groups from rule store", "userID", userID)
```

Comment:
These two log messages are the same, except first one tries to log *missing* namespace, while second one doesn't log namespace at all (even though it actually can exist there).

I suggest replacing second log with first one, and remove first one.

```suggestion
	}

		level.Debug(logger).Log("msg", "retrieving rule groups from rule store", "userID", userID, "namespace", namespace)
```

Reply:
👍